darwin/future/core/client.py

```diff
@@ -0,0 +1,203 @@
+    def __next__(self) -> Page:
+        pass
+
+
+class Client:
+    """Client Object to manage and make requests to the Darwin API
+    Attributes
+    ----------
+    url: str, url of the endpoint
+    api_key: str, api key to authenticate
+    team: Team, team to make requests to
+    """
+
+    def __init__(self, config: Config, retries: Optional[Retry] = None) -> None:
+        self.config = config
+        self.session = requests.Session()
+        if not retries:
+            retries = Retry(total=3, backoff_factor=0.2, status_forcelist=[500, 502, 503, 504])
+        self._setup_session(retries)
+        self._mappings = {
+            "get": self.session.get,
+            "put": self.session.put,
+            "post": self.session.post,
+            "delete": self.session.delete,
+            "patch": self.session.patch,
+        }
+
+    def _setup_session(self, retries: Retry) -> None:
+        self.session.headers.update(self.headers)
+        self.session.mount("http://", HTTPAdapter(max_retries=retries))
+        self.session.mount("https://", HTTPAdapter(max_retries=retries))
+
+    @property
+    def headers(self) -> Dict[str, str]:
+        http_headers: Dict[str, str] = {"Content-Type": "application/json"}
+        if self.config.api_key:
+            http_headers["Authorization"] = f"ApiKey {self.config.api_key}"
+        return http_headers
+
+    @overload
+    def _generic_call(self, method: Callable[[str], requests.Response], endpoint: str) -> dict:
+        ...
+
+    @overload
+    def _generic_call(self, method: Callable[[str, dict], requests.Response], endpoint: str, payload: dict) -> dict:
+        ...
+
+    def _generic_call(self, method: Callable, endpoint: str, payload: Optional[dict] = None) -> dict:
+        endpoint = self._validate_endpoint(endpoint)
+        url = self.config.base_url + endpoint
+        if payload is not None:
+            response = method(url, payload)
+        else:
+            response = method(url)
+
+        raise_for_darwin_exception(response)
+        response.raise_for_status()
+
+        return response.json()
+
+    def cursor(self) -> Cursor:
+        pass
+
+    def get(self, endpoint: str) -> dict:
+        return self._generic_call(self.session.get, endpoint)
+
+    def put(self, endpoint: str, data: dict) -> dict:
+        return self._generic_call(self.session.put, endpoint, data)
+
+    def post(self, endpoint: str, data: dict) -> dict:
+        return self._generic_call(self.session.post, endpoint, data)
+
+    def delete(self, endpoint: str) -> dict:
+        return self._generic_call(self.session.delete, endpoint)
+
+    def patch(self, endpoint: str, data: dict) -> dict:
+        return self._generic_call(self.session.patch, endpoint, data)
+
+    def _validate_endpoint(self, endpoint: str) -> str:
+        return endpoint.strip().strip("/")
+
+
+def raise_for_darwin_exception(response: requests.Response) -> None:
+    """Raises an exception if the response is not 200
```

Comment:
This doesn't raise if the error is in a 500 range, or 318 etc.?  Is that deliberate?

Reply:
It's a good catch on the documentation, it should instead state that it raises for known darwin exceptions rather than any non-200 code. The actual client function that uses it raises_for_status() after checking darwin codes, it'll catch every other code we don't have specific exceptions listed for.

Reply:
👍🏻